javascript/ql/src/semmle/javascript/NullSensitiveContext.qll

```diff
@@ -0,0 +1,488 @@
+/**
+ * Provides classes for working with contexts sensitive to containing null
+ * and undefined.
+ */
+
+import javascript
+
+/**
+ * A `NullSensitiveContext` is a syntactic context in which neither null nor
+ * undefined should occur, because the context is guaranteed to inspect its
+ * content and thereby throw an error or produce other undesirable behavior.
+ *
+ * The expression found in that context is called the principal argument of
+ * the context.
+ *
+ * In examples, we denote the position of the principal argument by use of
+ * an underscore. For example, in `_ + 0`, the context's principle argument
+ * is the left operand of the addition.
+ */
+abstract class NullSensitiveContext extends Expr {
+  abstract Expr getPrincipalArgument();
+}
+
+/*
+ * _[p]
+ */
+
+class IndexExprBase extends NullSensitiveContext, IndexExpr {
+  override Expr getPrincipalArgument() { result = this.getBase() }
+}
+
+/*
+ * o[_]
+ */
+
+class IndexExprIndex extends NullSensitiveContext, IndexExpr {
+  override Expr getPrincipalArgument() { result = this.getPropertyNameExpr() }
+}
+
+/*
+ * _.p
+ */
+
+class DotExprBase extends NullSensitiveContext, DotExpr {
+  override Expr getPrincipalArgument() { result = this.getBase() }
+}
+
+/*
+ * new _
+ */
+
+class NewExprCallee extends NullSensitiveContext, NewExpr {
+  override Expr getPrincipalArgument() { result = this.getCallee() }
+}
+
+/*
+ * _(args)
+ */
+
+class CallExprCallee extends NullSensitiveContext, CallExpr {
```

Comment:
You can unify this with the above by generalising from `NewExpr` to `InvokeExpr`.